### ISSUE TITLE
Thimble not loading (indexeddb, 3rd party cookies, Firefox)

### ISSUE THREAD

**Issue by humphd** [mozilla/thimble.mozilla.org#2427]
From a user, some fantastic debugging:

TL;DR I found the cause for the error: 3rd party cookie settings MUST be Always Allow.

Long story ahead… including one question why cookie site exceptions do not work.

The IndexedDB test fully passes:

Test for broken IndexedDB
This page tests the current browser's IndexexDB to see if it contains bugs.
Possible bugs you might see:

    IndexedDB not supported at all
    Samsung/HTC bug where IndexedDB doesn't have onupgradeneeded
    Apple bug in UIWebView on iOS 8 where IDBKeyRange is null rather than undefined
    Apple bug in desktop Safari 7.1+ and iOS 8+ Safari where IndexedDB doesn't allow opening two databases at once and is basically unusable. 

This browser has a working implementation of IndexedDB.

I also tried…

Disabling HTTPS Everywhere (active for Google APIs and Mozzilla.org)
 - did NOT work, no changes

Start Firefox with a new, clean profile which I created
 - did work, BUT, in the preview area the following 404 error was shown instead of the preview:

404 Not Found

    Code: NoSuchKey
    Message: The specified key does not exist.
    Key: bramble/production/dist/vfs/.anonymous/projects/ade1a7a7-5078-44c2-9b5e-c5353c3ead1f/index.html
    RequestId: 4ED40C443695904A
    HostId: gADo2unH/ici1uPeT8i25vcESOk1yPIby35zqyfyPIvUBhvxWbcFRPRIroNDjIDjg4H55RyKc98= 

An Error Occurred While Attempting to Retrieve a Custom Error Document

    Code: NoSuchKey
    Message: The specified key does not exist.
    Key: error.html 

After clicking the Refresh / reload button on top of the preview area displayed the preview correctly.
Race condition on an overloaded server? Anyway...

Further analysis
- In the inspection console the first (non fatal) warnings are displayed:

 Content Security Policy: Directive ‘frame-src’ has been deprecated. Please use directive ‘child-src’ instead.  (unknown)
The page was reloaded, because the character encoding declaration of the HTML document was not found when prescanning the first 1024 bytes of the file. The encoding declaration needs to be moved to be within the first 1024 bytes of the file.  index.html:27
The character encoding of a framed document was not declared. The document may appear different if viewed without the document framing it.  fa402bc8-582a-ae4f-994f-a184d988f299
The character encoding of a framed document was not declared. The document may appear different if viewed without the document framing it.  index.html

Finally tweaked settings in my new profile and got the cause:

in Privacy / Remember Cookies / Third party Cookies
- must be set to ALWAYS, From Visited or Never does not work.

Next experiment: add all cookies which are associated with Thimble to the allowed exception list.
- this does still NOT work when 3rd party cookie setting is set to From Visited or Never

Final experiment:
- set 3rd party cookie settings to ALWAYS alllow in default profile. This DOES work (but I don’t like this solution too much ;-)

**Comment by gideonthomas** [comment]
Do we know why Indexeddb is not accessible with cookie settings set to `From Visited`? It seems like an odd bug.

**Comment by humphd** [comment]
This was fixed by https://github.com/mozilla/brackets/pull/894.  See http://blog.humphd.org/promise-to-catch/ for more details.
